terminate/terminate.go

```diff
@@ -0,0 +1,209 @@
+
+	sampled := SampleSectors(sectors, PERC_SECTORS_TO_SAMPLE, MAX_SAMPLED_SECTORS, MIN_SAMPLED_SECTORS)
+
+	log.Println("Sampled sectors length: ", len(sampled))
+
+	// Sample 1% of sectors, with a max of 10000 and a minimum of 1000
+	sample := bitfield.NewFromSet(sampled)
+
+	res, err := TerminateSectors(ctx, api, minerAddr, &sample, ts)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	log.Printf("Termination penalty: %0.02f\n", util.ToFIL(res.TerminationFeeFromSample))
+
+	return res.EstimatedTotalTerminationFee, res.AvgTerminationFeePerPledge, nil
+}
+
+func AllSectors(ctx context.Context, api *lotusapi.FullNodeStruct, minerAddr address.Address, ts *types.TipSet) ([]uint64, error) {
+	allLiveSectors := make([]bitfield.BitField, 0)
+
+	for dlIdx := uint64(0); dlIdx < 48; dlIdx++ {
+		// 48 calls to the internet
+		partitions, err := api.StateMinerPartitions(ctx, minerAddr, uint64(dlIdx), ts.Key())
+		// if this error occurs, the miner hasn't been around long enough to process the deadlineTs, so we just skip it
+		if err != nil {
+			continue
+		}
+
+		for _, partition := range partitions {
+			allLiveSectors = append(allLiveSectors, partition.LiveSectors)
+		}
+	}
+
+	// one bitfield with all the sectors
+	allSectors, err := bitfield.MultiMerge(allLiveSectors...)
+	if err != nil {
+		return nil, err
+	}
+	// sectors => all sectors as an array of uint64 sector IDs
+	return allSectors.All(math.MaxUint64)
+}
+
+func SampleSectors(sectors []uint64, perc *big.Rat, max *big.Int, min *big.Int) []uint64 {
+	sectorCount := len(sectors)
+	if sectorCount == 0 {
+		return []uint64{}
+	}
+
+	// Compute the percentage of sectors
+	percFloat, _ := perc.Float64()
+	desiredCount := int(percFloat * float64(sectorCount))
+
+	// Apply the min and max boundaries
+	if big.NewInt(int64(desiredCount)).Cmp(min) < 0 {
+		desiredCount = int(min.Int64())
+	}
+	if big.NewInt(int64(desiredCount)).Cmp(max) > 0 {
+		desiredCount = int(max.Int64())
+	}
+
+	// Compute the step size for evenly spaced sampling
+	step := float64(sectorCount-1) / float64(desiredCount-1)
+
+	// Sample the sectors
+	sampledSectors := make([]uint64, desiredCount)
+	for i := 0; i < desiredCount; i++ {
+		index := int(float64(i) * step)
+		sampledSectors[i] = sectors[index]
+	}
+
+	return sampledSectors
+}
+
+func TerminateSectors(ctx context.Context, api *lotusapi.FullNodeStruct, minerAddr address.Address, sectors *bitfield.BitField, ts *types.TipSet) (*TerminateSectorResult, error) {
+	// create a new terminate sector result
+	result := TerminateSectorResult{
+		InitialPledgeFromSample:      big.NewInt(0),
+		TerminationFeeFromSample:     big.NewInt(0),
+		AvgInitialPledgePerSector:    big.NewInt(0),
+		AvgTerminationFeePerPledge:   big.NewInt(0),
+		EstimatedTotalInitialPledge:  big.NewInt(0),
+		EstimatedTotalTerminationFee: big.NewInt(0),
+		InitialPledge:                big.NewInt(0),
+		SectorCount:                  0,
+	}
+
+	_, mstate, err := mstat.LoadMinerActor(ctx, api, minerAddr, ts)
```

Comment:
i believe when we call this method, we potentially already have the MinerActor, so could maybe save a call to lotus here by changing the call params? up to you

Reply:
I left it in there